contracts/ExampleToken.cdc

```diff
@@ -44,228 +81,143 @@ pub contract ExampleToken: FungibleToken {
     /// out of thin air. A special Minter resource needs to be defined to mint
     /// new tokens.
     ///
-    pub resource Vault: FungibleToken.Provider, FungibleToken.Receiver, FungibleToken.Balance, MetadataViews.Resolver {
+    access(all) resource Vault: FungibleToken.Vault {
 
         /// The total balance of this vault
-        pub var balance: UFix64
+        access(all) var balance: UFix64
 
-        /// Initialize the balance at resource creation time
+        // initialize the balance at resource creation time
         init(balance: UFix64) {
             self.balance = balance
         }
 
+        /// Called when a fungible token is burned via the `Burner.burn()` method
+        access(contract) fun burnCallback() {
+            if self.balance > 0.0 {
+                ExampleToken.totalSupply = ExampleToken.totalSupply - self.balance
+            }
+            self.balance = 0.0
+        }
+
+        access(all) view fun getViews(): [Type] {
+            return ExampleToken.getContractViews(resourceType: nil)
+        }
+
+        access(all) fun resolveView(_ view: Type): AnyStruct? {
+            return ExampleToken.resolveContractView(resourceType: nil, viewType: view)
+        }
+
+        /// getSupportedVaultTypes optionally returns a list of vault types that this receiver accepts
+        access(all) view fun getSupportedVaultTypes(): {Type: Bool} {
+            let supportedTypes: {Type: Bool} = {}
+            supportedTypes[self.getType()] = true
```

Comment:
Out of curiosity, why isn't this function returning an array of `Type`? My guess is because each time we have to iterate to check if a certain type is supported, while a dictionary can be checked faster. The `trure` values are a side-effect of this being a dictionary.

Reply:
yeah, it is so people don't have to iterate